src/Uno.UI/UI/Xaml/PropertyMetadata.cs

```diff
@@ -122,11 +127,20 @@ BackingFieldUpdateCallback backingFieldUpdateCallback
 
 		public object DefaultValue
 		{
-			get => _defaultValue;
+			get
+			{
+				if (CreateDefaultValueCallback != null)
+				{
+					return CreateDefaultValueCallback.Invoke();
+				}
```

Comment:
Make sure to [validate the behavior](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.createdefaultvaluecallback?view=winrt-20348#remarks) of `CreateDefaultValueCallback`. We would definitely want to be behave like you implemented it, but that's unfortunately not the case. 

If I remember correctly, it's only called once per UI thread.

/cc @dr1rrb

Reply:
@dr1rrb Would you be able to provide help on how to do that?

Reply:
@Youssef1313 Can you validate the behavior first on UWP to see if the invocations are done as per what the documentation explain.

Reply:
@Youssef1313 the entry point of you work will require usage of the [`[ThreadStatic]` attribute](https://docs.microsoft.com/en-us/dotnet/api/system.threadstaticattribute?view=net-5.0) or it's more recent alternative a [`ThreadLocal`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.threadlocal-1?view=net-5.0)

Reply:
I started looking at this again, and I think I need to more understand the behavior on UWP before jumping into how to implement.

---

> If I remember correctly, it's only called once per UI thread.

I tried the following in UWP:

```csharp
    public class MyCustomButton : Button
    {

        public int MyCustomProperty
        {
            get { return (int)GetValue(MyCustomPropertyProperty); }
            set { SetValue(MyCustomPropertyProperty, value); }
        }

        // Using a DependencyProperty as the backing store for MyCustomProperty.  This enables animation, styling, binding, etc...
        public static readonly DependencyProperty MyCustomPropertyProperty =
            DependencyProperty.Register("MyCustomProperty", typeof(int), typeof(MyCustomButton), PropertyMetadata.Create(createDefaultValueCallback: GetValue));

        private static object GetValue()
        {
            return 50;
        }
    }
```

And in a single view app, I added a normal button to MainPage.xaml, and on click, I do:

```csharp
            var myCustomBtn = new MyCustomButton();
            _ = myCustomBtn.MyCustomProperty;

```

It seems that every click executes `GetValue`. ie, it's not once per thread.

---

I'll try to dig more and see if I can fully understand how this works and how to implement it.

Reply:
Oh I was creating new button each click, so it's very expected for it to execute every time 😄